A34-weighted-round-robin.md

```diff
@@ -1,103 +1,101 @@
 
 ### Overview of `weighted_round_robin` policy
+`weighted_round_robin` distribute traffic to each endpoint in a way that each endpoint will get fraction traffic equal to the weight associated with the endpoint divided by the sum of the weight of all endpoints. The core of `weighted_round_robin` is [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) picker.
+
+#### Weight of each endpoint
+`weighted_round_robin` needs extra information `weight` of each endpoint. We pass weight to `lb_policy` as per address attribute.
 
-`weighted_round_robin` policy is powered by [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) picker. EDF picker maintains a priority queue of `EdfEntry`. On the top of the queue, it’s the entry with lowest `deadline`. `order_offset` is the tie breaker when two entries have same deadline to maintain FIFO order.
+#### Overview of EDF scheduler
+EDF picker maintains a priority queue of `EdfEntry`. The key of the priority queue is `(deadline, order_offset)` pair. On the top of the queue, it’s the entry with lowest `deadline`. `order_offset` is the tie breaker when two entries have same deadline to maintain FIFO order. If there is a tie on deadline of two entries, the one with smaller `order_offset` will have higher priority.
 
 Proposed `EdfEntry`
 ```
 struct EdfEntry {
     // primary key for the priority queue. The entry with least deadline is the top of the queue.
     double deadline;
-    // secondary key for the priority queue. Used as a tiebreaker for same deadline to maintain FIFO order.
+
+    // secondary key for the priority queue. Used as a tiebreaker for same deadline to
+    // maintain FIFO order. If there is a tie on deadline of two entries, the one with
+    // smaller `order_offset` will have higher priority. `order_offset` is assigned to this
+    // entry on constructing the priority queue for the first time and it's immutable.
+    // Also the `order_offset` assigned to entries is strictly increasing, in other words,
+    // no two entries have same `order_offset`.
     uint64 order_offset;
-    // `load_balancing_weight` of this endpoint from the latest `ClusterLoadAssignment`.
+
+    // `load_balancing_weight` of this endpoint from address attribute of this endpoint.
     double weight;
```

Comment:
Why is this a `double` instead of a `uint32_t`?  The endpoint weight in the xDS API is an integer, and integers are usually a lot less error-prone to work with, because there's no rounding error.

Same question for the `deadline` field.  Is there a way we can represent is as an integer instead of a double?  For example, can we simply multiply by 1000 and say that we are storing milli-units?

Reply:
I make weight double because of deadline is double. I am not sure we can use `uint32` for deadline.  On updating deadline of an entry after a pick, one way to implement (or at least envoy does)
```
new_deadline = deadline + 1/weight
```
if we want to use `uint32`, we likely need to 
```
new_deadline = deadline + 1000/weight
```
or similar. This will need the assumption that per endpoint weight is never larger than 1000.

Reply:
If the deadlines and weights are all multiplied by 1000, then there's no requirement that endpoint weight is never larger than 1000.  We are doing exactly the same calculation as we would be doing with doubles, except that we're storing milliunits instead of whole units.  The only thing we're losing is the ability to store more 3 decimal places of precision, and that seems like it shouldn't really affect the algorithm in a material way.

Reply:
hmmm, sry, I am lost. how to handle the inverse of weight `1/weight`, anything larger than the numerator is rounded to 0?

Reply:
Okay, sorry, let me try this again.

The idea here is that you can do the actual division using a double, just don't use doubles for storage.  I think we don't actually need to store the weight as a double at all; it comes as an integer in xDS, and we can keep it that way.  We can then do the calculation as a double but immediately round the result for storage.

I think this can be done using something like this:

```
int32_t weight = <weight from xDS>;
int32_t deadline += std::lround(1000.0 / weight);
```

Then we continue to use the lowest deadline value as the sorting key.

Does that seem reasonable, or is there something I'm missing about this algorithm?  The latter is quite possible, since I haven't grokked it fully -- for example, if deadline is constantly increasing, how does it avoid wrap-around?  Is it depending on the `double` representation to avoid that?

Reply:
ah, I got the point, and it totally makes sense. I tried to avoid type casting, but yeah, we can avoid storing double. 

I don't remember there is any handling of wrap around out of head, so it depends on the double representation, and I think we need uint64 to stand against wrap around.

uint32 ~ `4*10^9 / 1000 ~ 4*10^6` requests
uint64 ~ `1.8*10^19 / 1000 ~ 1.8*10^16` requests

Reply:
We should probably figure out how the wrap around is handled, just to make sure we understand how this is supposed to work.  Can you figure out how Envoy does that and add info about it to this gRFC?

Reply:
ugh, Monday is not suitable for brainstorming work. 

```
std::lround(1000.0 / weight);
```
still goes to 0 if `weight` > 1000. I am sorry I keep back and forth on this. The issue I am confused is when `weight` is larger we need more precision in `deadline`.

re wrapping deadline:
per chat with @htuch:
it takes forever for double deadline to overflow, so no logic to handle it. Same argument can apply to uint64 i think.

Also, @htuch mentioned that envoy is reconsidering the use of EDF.
It's hard to avoid all grpc clients perform synchronized pick.
I will add this to open issues

Reply:
I see, so the problem is not the multiplying by 1000; it's the fact that the algorithm is actually counting on way more than 3 decimal digits' worth of precision.  Okay, then I guess we have no choice but to use a double here.  That's unfortunate, but I guess the use of it is fairly constrained here, so it shouldn't cause too many bugs.

What is Envoy considering using instead of EDF?  It seems like it would be better to not implement something that we'll immediately move away from.

Reply:
@htuch do you have timeline if envoy is going to discard EDF?

Reply:
@tonya11en is currently investigating alternatives, the current leading candidate is IWRR. The motivation is to provide a robust solution around desyncing, which is something that came up in work done by @mattklein123 and @snowp. See https://github.com/tonya11en/envoy/pull/1 for some early investigation, there may be other performance advantages of IWRR as well.

Reply:
If Envoy will have a decision soon, say this Q, let's put this PR on hold?

Reply:
I think we'll know in the next few weeks.

Reply:
I'm planning to make a more rigorous simulation over the weekend. Hopefully something we can just check in and use for other LBs. Main focus is on what the hosts in the pool experience w.r.t. syncing behavior. The perf numbers are more favoriable to IWRR, so we'll just need to see if there's a catch or unwanted change in behavior.

Stay tuned.

Reply:
Howdy folks! Happy New Year

@tonya11en do we have a final decision that envoy will use iwrr to replace edf?

Reply:
It looks like we'll be replacing EDF with WRSQ (in review https://github.com/envoyproxy/envoy/pull/14681) for weighted RR among other things. More details on where it'll be used in https://github.com/envoyproxy/envoy/issues/14597.

Reply:
Gotcha, thanks